### PR TITLE
Resolve Maven's `dependencyManagement` imports only once

### PR DESCRIPTION
**Pull request by timtebeek**
- Fixes https://github.com/openrewrite/rewrite/issues/2504

### CODE REVIEW

**Comment by shanman190** [comment]
So I don't think that we can do this. The reason here is that parents have higher precedence than BOMs. So if you were to use `mockito.version=4.3.0` with Maven, as an example, the managed dependencies found in the effective POM are 4.3.0 versions. With this change here, we'd use 4.6.1 erroneously.

File: rewrite-maven/src/main/java/org/openrewrite/maven/tree/ResolvedPom.java [hunk 1]
```diff
@@ -904,14 +907,18 @@ private GroupArtifactClassifierType createDependencyManagementKey(Defined define
...
+                Set<GroupArtifact> importedBoms) throws MavenDownloadingException {
             Pom pom = pomAncestry.get(0);
             for (ManagedDependency d : incomingDependencyManagement) {
                 if (d instanceof Imported) {
                     GroupArtifactVersion groupArtifactVersion = getValues(((Imported) d).getGav());
-                    if (isAlreadyResolved(groupArtifactVersion, pomAncestry)) {
+                    GroupArtifact bomGa = new GroupArtifact(groupArtifactVersion.getGroupId(), groupArtifactVersion.getArtifactId());
+                    // Skip if this BOM (by groupId:artifactId) was already imported or declared - first definition wins
+                    if (importedBoms.contains(bomGa) || isAlreadyResolved(groupArtifactVersion, pomAncestry)) {
                         continue;
```
Comment (sambsnyd) [inline rewrite-maven/src/main/java/org/openrewrite/maven/tree/ResolvedPom.java:919]: Maven's configuration model tends to start with the parent pom, then layer on child poms from there. This naturally adds up to "last wins" for most things.
We did not fully understand all the nuances when we implemented our Maven emulation originally, so we don't always do things in the same order, leading to edge cases like this popping up.

From maven's perspective the dependency management in the current pom is applied after the dependency management from the parent and last wins. But since we do some stuff a bit backwards it looks like first wins.

I bet if you had multiple dependency management definitions at the same level the semantics would come out looking more last-wins.

Anyway...fixing the order we parse things to more closely match Maven is a bigger refactoring than this fix. So if you have validated that our output matches `mvn dependency:tree` for the provided test case that could be good enough for now.

**Review by sambsnyd — COMMENTED** [review]
I'm a bit concerned that this might inadvertently take us further from maven's model rather than closer to it, see comment in the sources. But if you've verified that our model exactly matches `mvn dependency:tree` output for the test case maybe that's okay for now.

File: rewrite-maven/src/main/java/org/openrewrite/maven/tree/ResolvedPom.java (hunk 1 above)
Comment (nmck257) [reply rewrite-maven/src/main/java/org/openrewrite/maven/tree/ResolvedPom.java:919]: hmm -- I ran real-maven against the sample pom in this test case, and it aligns with what Shannon said earlier. Maven actually *does* similarly fail looking for the non-existent `org.mockito:mockito-bom:3.9.0`.

so, I think the test case I started with in #2504 must not have faithfully matched whatever I was seeing at the time. I think the best choice is to close the issue unless and until I encounter this mythical edge case again in the wild and can scrape out more specifics

thanks, all, for all the effort on this
